Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs

```diff
@@ -0,0 +1,308 @@
+
+        UpdateState(jukeBox, jukeboxComp);
+    }
+
+    #endregion Public functions
+    #region Private functions
+
+    /// <summary>
+    ///     Sets the Jukebox to either a state where it can or cannot play songs.
+    /// </summary>
+    private void CheckCanPlay(EntityUid uid, JukeboxComponent component)
+    {
+        var canPlay = true;
+
+        if (EntityManager.TryGetComponent<ApcPowerReceiverComponent>(uid, out var powerComp) && !powerComp.Powered)
+            canPlay = false;
+
+        component.CanPlay = canPlay;
+
+        if (canPlay)
+            TryRestart(uid, component);
+        else
+            Stop(uid, component);
+    }
+
+    /// <summary>
+    ///     Stops the currently playing song and sets the Jukebox to a stopped state.
+    /// </summary>
+    /// <remarks>
+    ///     See <see cref="Pause"/> to pause the Jukebox instead.
+    /// </remarks>
+    private void Stop(EntityUid jukeBox, JukeboxComponent jukeboxComp)
+    {
+        if (!jukeboxComp.Playing)
+            return;
+
+        jukeboxComp.Playing = false;
+
+        jukeboxComp.StoppedTime = _timing.CurTime;
+
+        if (jukeboxComp.CurrentlyPlayingStream != null)
+        {
+            jukeboxComp.CurrentlyPlayingStream.Stop();
+            jukeboxComp.CurrentlyPlayingStream = null;
+        }
+
+        jukeboxComp.SongTimerCancel.Cancel();
+
+        UpdateState(jukeBox, jukeboxComp);
+    }
+
+    /// <summary>
+    ///     Ends the currently playing song in the Jukebox.
+    /// </summary>
+    private void End(EntityUid jukeBox, JukeboxComponent jukeboxComp)
+    {
+        jukeboxComp.CurrentlyPlayingTrack = null;
+
+        jukeboxComp.FinishPlayingTime = null;
+
+        jukeboxComp.StoppedTime = null;
+
+        Clean(jukeBox, jukeboxComp);
+    }
+
+    /// <summary>
+    ///     Cleans up the active elements of a playing song.
+    /// </summary>
+    private void Clean(EntityUid jukeBox, JukeboxComponent jukeboxComp)
+    {
+        if (jukeboxComp.CurrentlyPlayingStream != null)
+        {
+            jukeboxComp.CurrentlyPlayingStream.Stop();
+            jukeboxComp.CurrentlyPlayingStream = null;
+        }
+
+        jukeboxComp.SongTimerCancel.Cancel();
+
+        UpdateState(jukeBox, jukeboxComp);
+    }
+
+    /// <summary>
+    ///     Tries to unpause the Jukebox and continue playing the current song from where it was paused.
+    /// </summary>
+    /// <returns>Whether or not the Jukebox is now playing.</returns>
+    private void TryRestart(EntityUid jukeBox, JukeboxComponent jukeboxComp)
+    {
+        if (jukeboxComp.Paused || !jukeboxComp.CanPlay)
+            return;
+
+        jukeboxComp.Playing = true;
+
+        if (jukeboxComp.CurrentlyPlayingTrack == null || jukeboxComp.FinishPlayingTime == null || jukeboxComp.StoppedTime == null)
+        {
+            UpdateState(jukeBox, jukeboxComp);
+            return;
+        }
```

Comment:
Should playing be true even if it isn't playing?... Seems like a dumb question, but I actually think that may be how this works...